### PR TITLE
Add OAuth docs

### PR DESCRIPTION
**Pull request by m-abdelwahab**
This PR adds the first version of the OAuth docs

Preview: https://neon-next-dnb4qk49l-neondatabase.vercel.app/docs/reference/oauth-integration

### CODE REVIEW

File: content/docs/reference/oauth-integration.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+- `grant_type`: This should be set to `authorization_code` to indicate that you are using the [Authorization Code grant type](https://oauth.net/2/grant-types/authorization-code/).
+- `code`: the authorization code returned from the previous step
+
+The response object will contain an `access_token` field value to send API calls on behalf of your users. You do that by passing it in the `Authorization` header of the HTTP request when sending requests to Neon's API.
+
+## Example apps
+
+You can check out the following example apps that leverage the Neon OAuth integration:
+
+- Visualizing Neon Postgres branches: [demo](https://neon-experimental.vercel.app), code(http//github.com/neon-bindings/neon-experimental)
```
Comment (duskpoet) [inline content/docs/reference/oauth-integration.md:80]: code link is broken

**Review by duskpoet — COMMENTED** [review]
Users don't actually need to provide all URLs manually, we support openid discoverability.
```
import { Issuer } from 'openid-client';

const neonIssuer = await Issuer.discover('https://oauth2.neon.tech');
```

Let's promote using openid client for integration https://www.passportjs.org/packages/openid-client/

File: content/docs/reference/oauth-integration.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+---
+title: Neon OAuth integration
+enableTableOfContents: true
+---
+
+
+You can integrate Neon into your application by leveraging our OAuth integration. The integration allows you to access your users' Neon accounts and act on their behalf.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:7]:
```suggestion
You can integrate your application with Neon by leveraging our secure OAuth integration. The integration enables your application to interact with users' Neon accounts, carrying out permitted actions on their behalf. Our integration does not require direct access to users' personal login credentials and is conducted with their  approval, ensuring data privacy and security.
```

File: content/docs/reference/oauth-integration.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+To set up the integration and create a Neon OAuth application, you can apply on our [Partners page](https://neon.tech/partner). You will need to provide us with the following information:
+
+1. Details about your application (name, what it does, link to its website)
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:11]:
```suggestion
1. Details about your application, such as its name, what it does, and a link to the website.
```

File: content/docs/reference/oauth-integration.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+2. Callback URL(s): URLs used to redirect users after completing the authorization flow. For example `https://yourapplication.com/api/oauth/callback`, `http://localhost:3000/api/oauth/callback`
+3. Scopes: scopes let you specify exactly what type of access you need. Right now, we provide access to the following scopes:
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:13]:
```suggestion
3. Scopes: scopes let you specify the type of access you require. Currently, we provide access to the following scopes:
```

File: content/docs/reference/oauth-integration.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+   - Create Projects
+   - Read Projects
+   - Modify Projects
+   - Delete Projects
+4. Whether or not you will be making API calls from a backend
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:18]:
```suggestion
4. Whether or not you will make API calls from a backend.
```

File: content/docs/reference/oauth-integration.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+5. Logo: an image that will be displayed on Neon's OAuth consent when users want to authorize your app to access their Neon account
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:19]:
```suggestion
5. Logo: an image to be displayed on Neon's OAuth consent dialog when users authorize your app to access their Neon account.
```

File: content/docs/reference/oauth-integration.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+After your application has been reviewed, we will get in touch with you and provide you with two credentials: a client ID and a client secret. These credentials are sensitive and should be stored securely.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:21]:
```suggestion
After your application is reviewed, we will get in touch with you and provide you with two credentials: a client ID and a client secret. These credentials are sensitive and should be stored securely.
```

File: content/docs/reference/oauth-integration.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+## How the OAuth integration works
+
+Here is a high-level overview of how Neon's OAuth implementatio works
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:25]:
```suggestion
Here is a high-level overview of how Neon's OAuth implementation works:
```

File: content/docs/reference/oauth-integration.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+![OAuth flow diagram](/docs/oauth/flow.png)
+
+1. The user initiates the OAuth flow in the third-party application by clicking a button or link.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:29]:
```suggestion
1. The user initiates the OAuth flow in your application by clicking a button or link.
```

File: content/docs/reference/oauth-integration.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+2. An authorization URL is generated, and the user is redirected to Neon’s OAuth consent screen, where they can authorize the third-party application and grant the necessary permissions.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:30]:
```suggestion
2. An authorization URL is generated, and the user is redirected to Neon’s OAuth consent screen, where they  authorize your application and grant the necessary permissions.
```

File: content/docs/reference/oauth-integration.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+3. Finally, the third-party application receives an access token to manage Neon resources on the user’s behalf.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:31]:
```suggestion
3. Finally, your application receives an access token to manage Neon resources on the user’s behalf.
```

File: content/docs/reference/oauth-integration.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+Here is a more detailed breakdown of the steps involved in the OAuth flow:
+
+### 1. Initiating the OAuth flow
+
+To initiate the OAuth flow, you will need to generate an authorization URL. To do that you can redirect your users to `https://oauth2.neon.tech/oauth2/auth` while passing the following query parameters:
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:37]:
```suggestion
To initiate the OAuth flow, you need to generate an authorization URL. You can do that by directing your users to `https://oauth2.neon.tech/oauth2/auth` while passing the following query parameters:
```

File: content/docs/reference/oauth-integration.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+- `client_id`: your OAuth application's ID
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:39]:
```suggestion
- `client_id`: your OAuth application's ID.
```

File: content/docs/reference/oauth-integration.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+- `redirect_uri`: the full URL that Neon should redirect users to after authorizing your app. This URL should match at least one of the callback URLs you provided when applying to become a partner.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:40]:
```suggestion
- `redirect_uri`: the full URL that Neon should redirect users to after authorizing your application. The URL should match at least one of the callback URLs you provided when applying to become a partner.
```

File: content/docs/reference/oauth-integration.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+- `scope`: This is a space-separated list of scopes that you want to request access to. Example value: `urn:neoncloud:projects:create urn:neoncloud:projects:read urn:neoncloud:projects:update urn:neoncloud:projects:delete`
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:41]:
```suggestion
- `scope`: This is a space-separated list of scopes that you want to request access to. For example: `urn:neoncloud:projects:create urn:neoncloud:projects:read urn:neoncloud:projects:update urn:neoncloud:projects:delete`
```

File: content/docs/reference/oauth-integration.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+- `response_type`: This should be set to `code` to indicate that you are using the [Authorization Code grant type]().
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:42]: missing link

File: content/docs/reference/oauth-integration.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+- `code_challenge`: This is a random string that will be used to verify the integrity of the authorization code.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:43]:
```suggestion
- `code_challenge`: This is a random string that is used to verify the integrity of the authorization code.
```

File: content/docs/reference/oauth-integration.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+- `state`: This is a random string that will be returned to your callback URL. You can use this parameter to verify that the request came from your application and not from a third party.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:44]:
```suggestion
- `state`: This is a random string that is returned to your callback URL. You can use this parameter to verify that the request came from your application and not from a third party.
```

File: content/docs/reference/oauth-integration.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+Here is an example of what the authorization URL might look like:
+
+```text
+https://oauth2.neon.tech/oauth2/auth?client_id=neon-experimental&scope=openid%20offline%20offline_access%20urn%3Aneoncloud%3Aprojects%3Acreate%20urn%3Aneoncloud%3Aprojects%3Aread%20urn%3Aneoncloud%3Aprojects%3Aupdate%20urn%3Aneoncloud%3Aprojects%3Adelete&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fcallback%2Fneon&grant_type=authorization_code&state=H58y-rSTebc3QmNbRjNTX9dL73-IyoU2T_WNievO9as&code_challenge=99XcbwOFU6iEsvXr77Xxwsk9I0GL4c4c4Q8yPIVrF_0&code_challenge_method=S256
+```
+
+After the user is redirected to the authorization URL, they will be presented with Neon's consent screen, which will be pre-populated with the scopes you requested. The user will be able to review the scopes and authorize your application to connect their Neon account.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:52]:
```suggestion
After the user is redirected to the authorization URL, they are presented with Neon's consent screen, which is pre-populated with the scopes you requested. From the consent screen, the user is able to review the scopes and authorize your application to connect their Neon account.
```

File: content/docs/reference/oauth-integration.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+![Neon OAuth consent screen](/docs/oauth/consent.png)
+
+### 2. Authorization code is returned to your callback URL
+
+After your user successfully completes the authorization flow, they will be redirected to the callback URL with the following query parameters appended to the URL:
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:58]:
```suggestion
After successfully completing the authorization flow, they user is redirected to the callback URL with the following query parameters appended to the URL:
```

File: content/docs/reference/oauth-integration.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+- `code`: an authorization code that will be exchanged for an access token
+- `scope`: the scopes that the user authorized your application to access
+- `state`: you can compare the value of this parameter with the original `state` you provided in the previous step to ensure that the request came from your application and not from a third party
+
+### 3. Exchanging the authorization code for an access token
+
+You can now exchange the authorization code returned from the previous step for an access token. To do that, you will need to send a `POST` request to `https://oauth2.neon.tech/oauth2/token` with the following parameters:
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:66]:
```suggestion
You can now exchange the authorization code returned from the previous step for an access token. To do that, you need to send a `POST` request to `https://oauth2.neon.tech/oauth2/token` with the following parameters:
```

File: content/docs/reference/oauth-integration.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+- `client_id`
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:68]: missing description

File: content/docs/reference/oauth-integration.md [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+- `redirect_uri`
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:69]: missing description

File: content/docs/reference/oauth-integration.md [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+- `client_secret`: your OAuth application's secret
+- `grant_type`: This should be set to `authorization_code` to indicate that you are using the [Authorization Code grant type](https://oauth.net/2/grant-types/authorization-code/).
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:71]:
```suggestion
- `grant_type`: set this to `authorization_code` to indicate that you are using the [Authorization Code grant type](https://oauth.net/2/grant-types/authorization-code/)
```

File: content/docs/reference/oauth-integration.md (hunk 22 above)
Comment (m-abdelwahab) [reply content/docs/reference/oauth-integration.md:68]: Should we repeat the description if it's the same as above?

File: content/docs/reference/oauth-integration.md [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+- `code`: the authorization code returned from the previous step
+
+The response object will contain an `access_token` field value to send API calls on behalf of your users. You do that by passing it in the `Authorization` header of the HTTP request when sending requests to Neon's API.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:74]:
```suggestion
The response object includes an `access_token` value, required for making requests to the Neon API on your users' behalf. This value must be supplied in the Authorization header of the HTTP request when sending requests to the Neon API.
```

File: content/docs/reference/oauth-integration.md [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+
+## Example apps
+
+You can check out the following example apps that leverage the Neon OAuth integration:
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:78]:
```suggestion
You can check out the following example applications that leverage the Neon OAuth integration:
```

**Comment by danieltprice** [comment]
Minor: For consistency, I suggest using "application" or "app", but not both. The term "application" seems a little more formal, which suits this context better, I think. If you go with "application", I suggest updating the diagram as well.

**Review by duskpoet — DISMISSED** [review]

File: content/docs/reference/oauth-integration.md [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+After your application is reviewed, we will get in touch with you and provide you with two credentials: a client ID and a client secret. These credentials are sensitive and should be stored securely.
+
+## How the OAuth integration works
+
+Here is a high-level overview of how Neon's OAuth implementation works:
+
+![OAuth flow diagram](/docs/oauth/flow.png)
+
+1. The user initiates the OAuth flow in your application by clicking a button or link.
+2. An authorization URL is generated, and the user is redirected to Neon’s OAuth consent screen, where they  authorize your application and grant the necessary permissions.
```
Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:30]:
```suggestion
2. An authorization URL is generated, and the user is redirected to Neon’s OAuth consent screen, where they  authorize your application and grant the necessary permissions.
```

Comment (danieltprice) [inline content/docs/reference/oauth-integration.md:30]: looks like an extra space here between they and authorize

**Review by danieltprice — APPROVED** [review]

**Pull request merged**